### PR TITLE
Upgrade codebase to C++ 20 + fix warnings + update submodules

### PR DESCRIPTION
**Pull request by GPUCode**
Does what it says on the tin. Updates most submodules to the latest versions, enables C++ 20 for the entire codebase and fixes some small warnings. In addition Link Time Optimization is added as an optional flag that is turned off by default. Another change is that SDL2 is built from source now, thus removing an extra prebuilt dependency and allowing to produce MinGW builds without an MSYS2 environment. Builds times are minimally affected since unsused components of SDL2 are disabled.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6115)

### CODE REVIEW

**Comment by ImperatorStorm** [comment]
I'll ask again, as I got no response on the previous PR

Is there an option to force system libraries for SDL in this PR? I don't see any immediately visible in the diff.

File: externals/sdl2/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+set(SDL_STATIC OFF CACHE BOOL "")
+
+# Subsystems
+set(SDL_ATOMIC ON CACHE BOOL "")
+set(SDL_AUDIO OFF CACHE BOOL "")
+set(SDL_VIDEO ON CACHE BOOL "")
+set(SDL_RENDER OFF CACHE BOOL "")
+set(SDL_EVENTS ON CACHE BOOL "")
+set(SDL_JOYSTICK ON CACHE BOOL "")
+set(SDL_HAPTIC ON CACHE BOOL "")
```
Comment (vitor-k) [inline externals/sdl2/CMakeLists.txt:39]: Does Citra use Haptic for anything? I remember looking into building sdl from source before and concluding it wasn't necessary (which is a difference from yuzu, as the switch has haptic feedback while afaik the 3ds doesn't).

**Comment by GPUCode** [comment]
> I'll ask again, as I got no response on the previous PR
> 
> Is there an option to force system libraries for SDL in this PR? I don't see any immediately visible in the diff.
> 
> I'd also like to bring up #5889:
> 
> > Currently, Citra bundles many dependencies on the repo. This is [widely considered a bad practice](https://fedoraproject.org/wiki/Bundled_Libraries) to be avoided as much as possible because it worsens security and is against many Linux distribution's [packaging policies](https://www.debian.org/doc/debian-policy/ch-source.html#s-embeddedfiles).

I did answer you in the previous PR you just didn't see it. Citra uses many different libraries many of which are not available in package managers and are linked statically. Currently only Qt, libusb and boost and can be provided from the system. This change to SDL2 was mostly for windows so it doesn't need prebuilts. If optional usage of system SDL2 is wanted I can implement it through

**Comment by SachinVin** [comment]
Please fix clang format too

File: src/common/logging/log.h [hunk 2]
```diff
@@ -1,13 +1,13 @@
-// Copyright 2014 Citra Emulator Project
+// Copyright 2022 Citra Emulator Project
```
Comment (SachinVin) [inline src/common/logging/log.h:1]: It's fine to leave it at 2014, it is the year when the file was created.

File: externals/sdl2/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+set(SDL_TIMERS ON CACHE BOOL "")
+set(SDL_FILE ON CACHE BOOL "")
+set(SDL_LOADSO ON CACHE BOOL "")
+set(SDL_CPUINFO OFF CACHE BOOL "")
+set(SDL_FILESYSTEM OFF CACHE BOOL "")
+set(SDL_DLOPEN ON CACHE BOOL "")
+set(SDL_SENSOR OFF CACHE BOOL "")
+set(SDL_LOCALE OFF CACHE BOOL "")
+
+add_subdirectory(SDL)
```
Comment (SachinVin) [inline externals/sdl2/CMakeLists.txt:52]: nit.. new line

File: src/common/logging/log.h (hunk 2 above)
Comment (GPUCode) [reply src/common/logging/log.h:1]: Shouldn't all copyright headers have an up-to-date date? I was planning to make a commit that does that eventually. Most projects I've seen keep update them though

Comment (SachinVin) [reply src/common/logging/log.h:1]: https://github.com/yuzu-emu/yuzu/pull/6853 here's an old pr from yuzu, there might have been an older and similar pr for Citra but it seems to have been buried. (It's not like copyright is going to expire anytime soon)

**Comment by mazes-80** [comment]
I don't think forcing to use bundled things is nice.
Especially for commonly distributed libraries as SDL2

**Comment by GPUCode** [comment]
> I don't think forcing to use bundled things is nice. Especially for commonly distributed libraries as SDL2

For linux I have no issue in checking for system libraries first and I will add an option for it before this PR is merged. This is mostly for windows though which has been living with an outdated version for ages. And as I mentioned it allows for MinGW builds without downloading the entire MSYS2 environment

**Comment by mazes-80** [comment]
> For linux I have no issue in checking for system libraries first and I will add an option for it before this PR is merged. This is mostly for windows though which has been living with an outdated version for ages. And as I mentioned it allows for MinGW builds without downloading the entire MSYS2 environment

thanks so much, I struggle to unbundle most libs I can when possible when I make a package.
As your PR already landed in canary I'm currently working on re-unbundling sdl2
and add an option to use default bundled catch as I unbundled it but stable system package is too old here

**Comment by GPUCode** [comment]
> > For linux I have no issue in checking for system libraries first and I will add an option for it before this PR is merged. This is mostly for windows though which has been living with an outdated version for ages. And as I mentioned it allows for MinGW builds without downloading the entire MSYS2 environment
> 
> thanks so much, I struggle to unbundle most libs I can when possible when I make a package. As your PR already landed in canary I'm currently working on re-unbundling sdl2 and add an option to use default bundled catch as I unbundled it but stable system package is too old here

Unbundling every library is not good both for performance and compatibility reasons. Especially statically linked library like Catch should be compiled from source so the compiler can properly optimize the library. Shared libraries like SDL, Qt, libusb are perfectly fine though

**Comment by ImperatorStorm** [comment]
> Especially statically linked library like Catch should be compiled from source so the compiler can properly optimize the library.

Isn't this already done for SO libs on linux?

**Comment by SachinVin** [comment]
untagged because it conflicts with #6122

File: externals/sdl2/CMakeLists.txt (hunk 1 above)
Comment (GPUCode) [reply externals/sdl2/CMakeLists.txt]: Disabled haptic

File: src/common/logging/log.h (hunk 2 above)
Comment (GPUCode) [reply src/common/logging/log.h]: I'll revert the data then

**Comment by mazes-80** [comment]
> > I don't think forcing to use bundled things is nice. Especially for commonly distributed libraries as SDL2
> 
> For linux I have no issue in checking for system libraries first and I will add an option for it before this PR is merged. This is mostly for windows though which has been living with an outdated version for ages. And as I mentioned it allows for MinGW builds without downloading the entire MSYS2 environment

Forgot to add an alias for SDL2 when enabling USE_SYSTEM_SDL2
find_package(SDL2 ...
if(SDL2_FOUND)
    add_library(SDL2 INTERFACE)
    target_link_libraries(SDL2 INTERFACE "${SDL2_LIBRARY}")
    target_include_directories(SDL2 INTERFACE "${SDL2_INCLUDE_DIR}")
    add_library(SDL2::SDL2 ALIAS SDL2)
endif()

**Comment by GPUCode** [comment]
> > > I don't think forcing to use bundled things is nice. Especially for commonly distributed libraries as SDL2
> > 
> > 
> > For linux I have no issue in checking for system libraries first and I will add an option for it before this PR is merged. This is mostly for windows though which has been living with an outdated version for ages. And as I mentioned it allows for MinGW builds without downloading the entire MSYS2 environment
> 
> Forgot to add an alias for SDL2 when enabling USE_SYSTEM_SDL2 `find_package(SDL2 ...` `if(SDL2_FOUND)` ` add_library(SDL2 INTERFACE)` ` target_link_libraries(SDL2 INTERFACE "${SDL2_LIBRARY}")` ` target_include_directories(SDL2 INTERFACE "${SDL2_INCLUDE_DIR}")` ` add_library(SDL2::SDL2 ALIAS SDL2)` `endif()`
> 
> linked to system sdl with no troubles so

Opened a PR to fix this https://github.com/citra-emu/citra/pull/6134

**Pull request merged**